Scope uninitialized

o- opened the pull request:
Initially I just wanted to fix a warning (see 336a468).

But then I realized that scope inference and checking for uninitialized variables really should be split into two independent steps

o- committed:
improve warnings for uninitialized variables

We should not confuse the check for uninitialized variables with the
notion of scopes.

when I write

     mut x
    loop:
     x <- (x+1)
     branch (x==10) end loop
    end:

I find the old warning:

    scope.sou:3 : Error: Variable x might be uninitialized.

A lot more clear than now:

    At instruction 1 [...] the latter defines {mut x} and the former does not

o- committed:
disentangle scope inference and the check for uninitialized variables

this prevents future harm and makes it simpler to read

o- committed:
deal with sometimes uninitialized variables

variables can be uninitialized only on some converging paths.
The old existing analysis was overly conservative!

o- committed:
add a positive example

o- commented:
Update: @gasche did you intend to not allow a merge between a undefined/defined variable?

Consider e.g.

```
  mut x = 1
  branch (x==1) clearit skip
clearit:
  clear x
skip:
  x <- 1
  print x
```

This was illegal at pc 4 since it merges x in defined and undefined state. Imho this is too conservative (also our scoping judments never forbid this) and the changes in 0d790c6 make it legal again

gasche commented:
This was indeed my intention: I thought that it would be consistent with the scoping rule, given that I also added an explicit `clear x` instruction to make mutable variable undefined again.

If you think that relaxing this constraint is better, we can go for it. It was my intuition, however, that the stricter constraint would also be useful for optimization purposes: knowing for sure that every parent defined the value seems like a good invariant to have. (It also has the convenient consequence that we completely rule out "undefined value" errors at runtime, instead of having just a heuristic warning.)

o- commented:
> This was indeed my intention

I see. I don't have a strong opinion to be honest. I just wanted the warning not to talk about scopes when the problem was that a variable is maybe undefined. Since it confused even me.

> It also has the convenient consequence that we completely rule out "undefined value" errors at runtime, instead of having just a heuristic warning.

That is not true. The proposed chage is less conservative. But it's still conservative. It won't allow use of an undefined variable.

o- commented:
I think I will revert to your intended behavior but keep the two analysis separate and improve the wording of the error message. How about that?

gasche commented:
> The proposed chage is less conservative. But it's still conservative. It won't allow use of an undefined variable.

Oh, I assumed you merged (Uninit + Init -> Init), of course if you merge (-> Uninit) then it's fine, sorry.

I am fine with a different error message. What are the benefits of splitting the two analyses, though? It creates some amount of code duplication. Could have the better message with a single-pass computation still? I'm just asking of curiosity, I have no strong opinion; the factoring of the two computations (and the Scope_info) type came from you anyway, so you are welcome to take it back if you prefer ;-)

o- commented:
> What are the benefits of splitting the two analyses, though?

It seems to create accidental complexity because the two things do not really belong together. There is not much code duplication since the common stuff is in forward_analysis. If I remove the scope_info code gets shorter (even with the duplicated lines). But I guess it's a matter of taste and I am not even sure myself :)

gasche commented:
I propose then to honor your work by merging the PR.

At some point we might want to have some form of integration tests that also check the error messages. (I like the fact that having this makes it easy to review error-message changes in PRs as they are present in full in reference files for tests.)

o- committed:
revert to the old behavior but with improved error message

after discussion with @gasche we want to keep the analysis more
conservative than necessary. It should not be allowed to have
a case where we are not sure if something is defined or undefined.

o- commented:
I am totally unsure whether the transformations in #58 respect those additional scoping rules! One argument against them would be that moving code around does have to respect two different scoping rules (actual scopes and uninitialized-scopes).

Ie. without this restrictions you can:

1. maximize scope
2. move code
3. shrink scope

this is what I more or less do in #58 (at least 1&2) by converting assignment into declaration.

With the restriction (2) has to respect the uninitialized-scopes. I have to think more about what this entails.

o- commented:
Lets assume the following example:

```
  mut a
  branch exp b1 b2
b1:
  a <- 1
  print a
  goto cont
b2:
  mut b = 1
  a <- b
  print a
cont:
```

Now if we perform the assignment hoisting from #58 the following will happen:

```
  mut a0 <- 1
  mut a
  branch exp b1 b2
b1:
  print a0
  goto cont
b2:
  mut b = 1
  a <- b
  print a
cont:
```

Conclusion: moving around code can invalidate your proposed invariant. Recovering it would be complicated.

Scopes are easier to preserve since they have only one starting point (ie. the declaration). But the undefined/defined can have arbitrarily many assignments to the same variable.

Therefore I'd like to lift this requirement for now.

o- committed:
revert scope definition again

o- commented:
superseeded by #60

gasche commented:
(Thinking out loud, this is not in contradiction to what you say.) If `a` was used after `cont` in your example, then the proposed rewriting would be wrong: you would need to assign `a <- a0` in `b1`. If `a` is not used, and `clear,drop` were maximally placed, then the issue would not occur: before both transitions to `cont` we would have a `drop` if `a` is not mentioned anymore, and a `clear` if `a` is assigned but not required.

o- commented:
> If a was used after cont in your example, then the proposed rewriting would be wrong:

agreed. In the parlance of #58 the assignment would not dominate all its uses (therefore no lifting will be done).

> If a is not used, and clear,drop were maximally placed, then the issue would not occur:

I think you might be right. Let's keep this in mind! when we have resolved the issue of how to place drops and clears maybe we can resurrect the invariant.
